determined-ai/determined
# fix: signed payload generation

## Description

The webhooks functionality has finally landed and testing has uncovered a few bugs that have been updated in this PR. 

**Updates**
1. Previously `generateSignedPayload` was not using the correct request body, this has been fixed.
2. Previously `generateSignedPayload` was not computing the correct hmac, this has been fixed.
3. `e2e tests` were failing (https://app.circleci.com/pipelines/github/determined-ai/determined/29244/workflows/686cd76b-1866-41f7-ae31-5181caed8864/jobs/993872/parallel-runs/2/steps/2-133) due to webhook events being received by the mock server during the test that were 
not the expected request. This has been fixed by only checking a subset of the expected payload.
4. The documents have been updated to note that users may need to double check the formatting of the 
request payload based on the language or library that they are using. 




## Test Plan

1. Create a Webhook with a trigger for a completed experiment. I recommend using a url from `webhook.site`.
2. Run an experiment that will complete. 
3. Copy the webhook request body, X-Determined-AI-Signature, X-Determined-AI-Signature-Timestamp, and the webhook signing key from the cluster. 
4. Run the following script here and replace the values with the information above and ensure that the signed payload matches: https://onecompiler.com/python/3ym8ryu32

Example:
![Screen Shot 2022-10-27 at 9 48 13 PM](https://user-images.githubusercontent.com/103522725/198479251-91e325b6-7c65-4ca0-a469-0d9a605f4eaa.png)
![Screen Shot 2022-10-27 at 9 48 38 PM](https://user-images.githubusercontent.com/103522725/198479433-19a34015-9cbf-401a-8957-ddfca386725a.png)





## Commentary (optional)




## Checklist
~- [ ] Release notes should be added as a separate file under `docs/release-notes/`.~
~See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.~
~- [ ] Licenses should be included for new code which was copied and/or modified from any external code.~
